- Refactor ember-debug to use ES6.

Mawaheb opened the pull request:
OK, I believe this is ready, But, there are a few problems that I faced:

- I couldn't refactor the contents of `/vendor`, when I tried, I got an error "Can't use block-scoped let, const out of restricted mode."

- Because arrow functions should always be anonymous, I couldn't/figure out how to refactor [this one](https://github.com/emberjs/ember-inspector/blob/master/ember_debug/view-debug.js#L192) to use `this` instead of `self`
similar problem with [this one](https://github.com/emberjs/ember-inspector/blob/master/ember_debug/deprecation-debug.js#L177)


As always, Any suggestions, comments, notes, things to be changed or feedback is always appreciated

teddyzeenny commented:
:+1: Thanks @Mawaheb! I will review this weekend at the latest.

Mawaheb commented:
No problem :) My pleasure.

Patsy-issa commented on ember_debug/adapters/basic.js:
```diff
@@ -1,15 +1,12 @@
-      self.onConnectionReady();
+const Ember = window.Ember;
+const { $, A, computed, RSVP, Object } = Ember;
+const { Promise } = RSVP;
```
`const { Promise, resolve } = RSVP;`

Patsy-issa commented on ember_debug/adapters/basic.js:
```diff
@@ -1,15 +1,12 @@
+
+export default Object.extend({
+  init() {
+    RSVP.resolve(this.connect(), 'ember-inspector').then(() => {
```
`resolve(...`

Patsy-issa commented on ember_debug/adapters/basic.js:
```diff
@@ -69,11 +66,11 @@ export default Ember.Object.extend({
+  _messageReceived(message) {
     this.get('_messageCallbacks').forEach(function(callback) {
       callback.call(null, message);
     });
```
```javascript
this.get('_messageCallbacks').forEach(callback => callback.call(null, message));
```

Patsy-issa commented on ember_debug/adapters/basic.js:
```diff
@@ -149,12 +145,11 @@ export default Ember.Object.extend({
+    const messages = this.get('_pendingMessages');
+    messages.forEach((options) => {
+      this.sendMessage(options);
     });
```
```javascript
messages.forEach(options => this.sendMessage(options));
```

Mawaheb commented:
Thanks for your feedback @Patsy-issa  :) .

Patsy-issa commented:
Np ^^ I ll take a look at the other files when I have the time.

teddyzeenny commented on ember_debug/adapters/basic.js:
```diff
@@ -1,15 +1,12 @@
-    RSVP.resolve(this.connect(), 'ember-inspector').then(function() {
-      self.onConnectionReady();
+const Ember = window.Ember;
+const { $, A, computed, RSVP, Object } = Ember;
```
`Object` has meaning in native javascript. You should rename it to something like `EmberObject`:
`const { $, A, computed, RSVP, Object: EmberObject } = Ember;`

teddyzeenny commented on ember_debug/adapters/chrome.js:
```diff
@@ -1,42 +1,41 @@
+const ChromeAdapter = BasicAdapter.extend({
+  connect() {
+    const channel = this.get('_channel');
+    return this._super.apply(this, arguments).then(() => {
```
ES6 spread operator: `this._super(...arguments)`

teddyzeenny commented on ember_debug/adapters/firefox.js:
```diff
@@ -1,14 +1,15 @@
 
-var FirefoxAdapter = BasicAdapter.extend({
-  init: function() {
+const FirefoxAdapter = BasicAdapter.extend({
```
Let's just export default instead of two steps.

teddyzeenny commented on ember_debug/adapters/websocket.js:
```diff
@@ -1,49 +1,45 @@
+const { $, computed, run, RSVP: { Promise } } = Ember;
 
-var WebsocketAdapter = BasicAdapter.extend({
+const WebsocketAdapter = BasicAdapter.extend({
```
`export default BasicAdapter.extend`

teddyzeenny commented on ember_debug/adapters/websocket.js:
```diff
@@ -1,49 +1,45 @@
-  connect: function() {
-    var self = this;
+  connect() {
     return new Promise(function(resolve, reject) {
```
`return new Promise((resolve, reject) => {`

teddyzeenny commented on ember_debug/addons/ember-new-computed/index.js:
```diff
@@ -1,21 +1,21 @@
-  var polyfillArguments = [];
-  var config = arguments[arguments.length - 1];
+  let polyfillArguments = [],
+      config = arguments[arguments.length - 1];
```
Single variable declaration with assignment (same as Ember core https://github.com/emberjs/ember.js/pull/10037).
```js
let polyfillArguments = [];
let config = arguments[arguments.length - 1];
```

teddyzeenny commented on ember_debug/container-debug.js:
```diff
@@ -1,19 +1,18 @@
-var computed = Ember.computed;
-var oneWay = computed.oneWay;
+const Ember = window.Ember;
+const { Object, computed } = Ember;
```
`const { Object: EmberObject, computed } = Ember;`

teddyzeenny commented on ember_debug/container-debug.js:
```diff
@@ -24,29 +23,29 @@ export default EmberObject.extend(PortMixin, {
-    var cache = this.get('container').cache;
+  instancesByType() {
+    let key, instancesByType = {},
+        cache = this.get('container').cache;
```
```js
let key;
let instancesByType = {};
let cache = this.get('container').cache;
```

teddyzeenny commented on ember_debug/container-debug.js:
```diff
@@ -59,37 +58,37 @@ export default EmberObject.extend(PortMixin, {
-    var key, types = [];
-    var instancesByType = this.instancesByType();
+  getTypes() {
+    let key, types = [];
```
```js
let key;
let types = [];
```

teddyzeenny commented on ember_debug/data-debug.js:
```diff
@@ -1,7 +1,10 @@
-var DataDebug = Ember.Object.extend(PortMixin, {
-  init: function() {
+const Ember = window.Ember;
+const { Object, computed, guidFor } = Ember;
```
`Object: EmberObject`

teddyzeenny commented on ember_debug/data-debug.js:
```diff
@@ -1,7 +1,10 @@
+const { Object, computed, guidFor } = Ember;
+const { alias } = computed;
+
+const DataDebug = Object.extend(PortMixin, {
```
`export default EmberObject.extend`

teddyzeenny commented on ember_debug/data-debug.js:
```diff
@@ -36,34 +39,33 @@ var DataDebug = Ember.Object.extend(PortMixin, {
+    let typesToSend;
+    typesToSend = types.map((type) => {
+      return this.wrapType(type);
     });
```
Tip: single-line arrow functions can have implicit returns if you don't use curly brackets:
```js
types.map(type => this.wrapType(type));
```

teddyzeenny commented on ember_debug/data-debug.js:
```diff
@@ -36,34 +39,33 @@ var DataDebug = Ember.Object.extend(PortMixin, {
+  modelTypesUpdated(types) {
+    let typesToSend = types.map((type) => {
+      return this.wrapType(type);
     });
```
`let typesToSend = types.map(type => this.wrapType(type));`

teddyzeenny commented on ember_debug/data-debug.js:
```diff
@@ -75,40 +77,39 @@ var DataDebug = Ember.Object.extend(PortMixin, {
+  wrapRecord(record) {
+    const objectId = guidFor(record.object);
+    let columnValues = {},
+        searchKeywords = [];
```
```js
let columnValues = {};
let searchKeywords = [];
```

teddyzeenny commented on ember_debug/data-debug.js:
```diff
@@ -140,67 +141,57 @@ var DataDebug = Ember.Object.extend(PortMixin, {
-        function() {
-          self.recordsRemoved.apply(self, arguments);
+        () => {
+          this.recordsRemoved.apply(this, arguments);
```
`this.recordsRemoved(...arguments);`

teddyzeenny commented on ember_debug/deprecation-debug.js:
```diff
@@ -1,14 +1,11 @@
-
-export default EmberObject.extend(PortMixin, {
+const Ember = window.Ember;
+const { Object, computed, guidFor, run, RSVP, A } = Ember;
```
`Object: EmberObject`

teddyzeenny commented on ember_debug/deprecation-debug.js:
```diff
@@ -59,21 +55,20 @@ export default EmberObject.extend(PortMixin, {
+  sendPending() {
+    let deprecations = A();
+
+    let promises = RSVP.all(this.get('deprecationsToSend').map((deprecation) => {
```
`let promises = all(`

teddyzeenny commented on ember_debug/deprecation-debug.js:
```diff
@@ -59,21 +55,20 @@ export default EmberObject.extend(PortMixin, {
+    let promises = RSVP.all(this.get('deprecationsToSend').map((deprecation) => {
+      let obj,
+          promise = resolve(undefined, 'ember-inspector'),
+          grouped = this.get('groupedDeprecations');
```
Single variable declaration with assignment.

teddyzeenny commented on ember_debug/deprecation-debug.js:
```diff
@@ -98,34 +93,34 @@ export default EmberObject.extend(PortMixin, {
-      self.sendMessage('deprecationsAdded', {
+    promises.then(() => {
+      this.sendMessage('deprecationsAdded', {
         deprecations: deprecations
```
`this.sendMessage('deprecationsAdded', { deprecations });`

teddyzeenny commented on ember_debug/deprecation-debug.js:
```diff
@@ -137,49 +132,49 @@ export default EmberObject.extend(PortMixin, {
+      let noDeprecation;
 
-      if (typeof test === 'function' && !(Ember.Object.detect(test))) {
+      if (typeof test === 'function' && !(Object.detect(test))) {
```
`if (typeof test === 'function' && !(EmberObject.detect(test))) {`

teddyzeenny commented on ember_debug/general-debug.js:
```diff
@@ -1,10 +1,10 @@
-var computed = Ember.computed;
-var oneWay = computed.oneWay;
+const Ember = window.Ember;
+const { $, computed, Object, A } = Ember;
```
`Object: EmberObject`

teddyzeenny commented on ember_debug/general-debug.js:
```diff
@@ -31,27 +30,27 @@ var GeneralDebug = Ember.Object.extend(PortMixin, {
-    getLibraries: function() {
-      var libraries = Ember.libraries;
+    getLibraries() {
+      let libraries = Ember.libraries;
```
Is it possible to import `libraries` as a const at the top of the file?
`const { libraries } = Ember`

teddyzeenny commented on ember_debug/general-debug.js:
```diff
@@ -62,14 +61,14 @@ var GeneralDebug = Ember.Object.extend(PortMixin, {
 
 function arrayize(enumerable) {
-  return Ember.A(enumerable).map(function(item) {
+  return A(enumerable).map(function(item) {
```
`return A(enumerable).map(item => item);`

teddyzeenny commented on ember_debug/libs/promise-assembler.js:
```diff
@@ -6,34 +6,34 @@
 import Promise from 'ember-debug/models/promise';
-var Ember = window.Ember;
+const Ember = window.Ember;
+const { Object, Evented, A, computed, RSVP, copy, isNone } = Ember;
```
`Object: EmberObject`

teddyzeenny commented on ember_debug/libs/promise-assembler.js:
```diff
@@ -6,34 +6,34 @@
+let PromiseAssembler = Object.extend(Evented, {
   // RSVP lib to debug
-  RSVP: Ember.RSVP,
+  RSVP: RSVP,
```
You can just use: 
```js
// RSVP lib to debug
RSVP,
```

teddyzeenny commented on ember_debug/libs/source-map.js:
```diff
@@ -3,15 +3,12 @@
-var RSVP = Ember.RSVP;
-var resolve = RSVP.resolve;
+const Ember = window.Ember;
+const { Object, A, computed, RSVP: { resolve, Promise } } = Ember;
```
`Object: EmberObject`

teddyzeenny commented on ember_debug/libs/source-map.js:
```diff
@@ -24,17 +21,16 @@ export default EmberObject.extend({
+  map(stack) {
+    const parsed = A(fromStackProperty(stack));
+    let array = A(),
+        lastPromise = null;
```
```js
let array = A();
let lastPromise = null;
```

teddyzeenny commented on ember_debug/libs/source-map.js:
```diff
@@ -57,13 +53,13 @@ export default EmberObject.extend({
+  getSourceMap(url) {
+    let sourceMaps = this.get('sourceMapCache');
     if (sourceMaps[url] !== undefined) { return resolve(sourceMaps[url], 'ember-inspector'); }
     return retrieveSourceMap(url).then(function(response) {
```
`.then(response => {`

teddyzeenny commented on ember_debug/main.js:
```diff
@@ -11,9 +11,10 @@ import ContainerDebug from "ember-debug/container-debug";
-var Ember = window.Ember;
-EmberDebug = Ember.Object.extend({
+const Ember = window.Ember;
+const { Object, run, Application, namespaces } = Ember;
```
`Object: EmberObject`

teddyzeenny commented on ember_debug/main.js:
```diff
@@ -52,27 +52,27 @@ EmberDebug = Ember.Object.extend({
+  reset($keepAdapter) {
     this.destroyContainer();
-    Ember.run(this, function() {
+    run(this, function() {
```
`run(() => {`

teddyzeenny commented on ember_debug/models/promise.js:
```diff
@@ -1,13 +1,14 @@
 import computedPolyfill from '../addons/ember-new-computed/index';
-var Ember = window.Ember;
+const Ember = window.Ember;
+const { typeOf, Object, computed, A } = Ember;
```
`Object: EmberObject`

teddyzeenny commented on ember_debug/object-inspector.js:
```diff
@@ -60,9 +51,9 @@ function inspect(value) {
-    var ret = [], v, count = 0, broken = false;
-    for (var key in value) {
-      if (!('hasOwnProperty' in value) || value.hasOwnProperty(key)) {
+    let ret = [], v, count = 0, broken = false;
```
```js
let v;
let ret = [];
let count = 0;
let broken = false;
```

teddyzeenny commented on ember_debug/object-inspector.js:
```diff
@@ -85,7 +76,7 @@ function inspect(value) {
 }
 
-var ObjectInspector = EmberObject.extend(PortMixin, {
+let ObjectInspector = EmberObject.extend(PortMixin, {
```
`export default EmberObject.extend(PortMixin, {`

teddyzeenny commented on ember_debug/object-inspector.js:
```diff
@@ -247,10 +237,9 @@ var ObjectInspector = EmberObject.extend(PortMixin, {
-        self = this;
+  retainObject(object) {
+    let meta = emberMeta(object),
+        guid = guidFor(object);
```
```js
let meta = emberMeta(object),
let guid = guidFor(object);
```

teddyzeenny commented on ember_debug/object-inspector.js:
```diff
@@ -380,17 +369,17 @@ var ObjectInspector = EmberObject.extend(PortMixin, {
-    var object = this.sentObjects[objectId];
+    let object = this.sentObjects[objectId];
 
     function handler() {
```
I think we can remove `self` if we use:
`const handler = () =>`

teddyzeenny commented on ember_debug/object-inspector.js:
```diff
@@ -475,18 +463,18 @@ function replaceProperty(properties, name, value, options) {
   if (found) { properties.splice(i, 1); }
 
-  var prop = { name: name, value: inspectValue(value) };
+  let prop = { name: name, value: inspectValue(value) };
```
`let prop = { name, value: inspectValue(value) };`

teddyzeenny commented on ember_debug/object-inspector.js:
```diff
@@ -707,7 +695,7 @@ function isComputed(value) {
-  return keys(errors).map(function(key) {
+  return keys(errors).map(key => {
     return errors[key];
   });
```
This can be:
```js
return keys(errors).map(key => errors[key])
```

teddyzeenny commented on ember_debug/port.js:
```diff
@@ -1,34 +1,32 @@
-var guidFor = Ember.guidFor;
-var run = Ember.run;
+const Ember = window.Ember;
+const { Object, computed, guidFor, run } = Ember;
```
`Object: EmberObject`

teddyzeenny commented on ember_debug/render-debug.js:
```diff
@@ -1,17 +1,14 @@
-var oneWay = computed.oneWay;
-var later = Ember.run.later;
+const Ember = window.Ember;
+const { computed: { oneWay }, run: { later }, subscribe, Object } = Ember;
```
`Object: EmberObject`

teddyzeenny commented on ember_debug/render-debug.js:
```diff
@@ -1,17 +1,14 @@
-
-var queue = [];
+let profileManager = new ProfileManager(),
+    queue = [];
```
```js
let profileManager = new ProfileManager(),
let queue = [];
```

teddyzeenny commented on ember_debug/render-debug.js:
```diff
@@ -67,18 +64,17 @@ export default Ember.Object.extend(PortMixin, {
+    this.profileManager.wrapForErrors = (context, callback) => {
+      return this.get('port').wrap(function() {
         return callback.call(context);
       });
```
`return this.get('port').wrap(() => callback.call(context))`

teddyzeenny commented on ember_debug/route-debug.js:
```diff
@@ -1,27 +1,23 @@
-
-export default Ember.Object.extend(PortMixin, {
+const Ember = window.Ember;
+const { String: { classify, dasherize }, computed, observer, run: { later }, Object } = Ember;
```
`Object: EmberObject`

teddyzeenny commented on ember_debug/route-debug.js:
```diff
@@ -30,50 +26,49 @@ export default Ember.Object.extend(PortMixin, {
-    later(function() {
-      self.sendMessage('currentRoute', { name: self.get('currentPath') });
+    later(() => {
+      this.sendMessage('currentRoute', { name: self.get('currentPath') });
```
`this.get('currentPath')`

teddyzeenny commented on ember_debug/route-debug.js:
```diff
@@ -30,50 +26,49 @@ export default Ember.Object.extend(PortMixin, {
     }
   },
 
   sendCurrentRoute: observer(function() {
```
`observer('currentPath', function() {`

teddyzeenny commented on ember_debug/services/session.js:
```diff
@@ -1,7 +1,7 @@
-var Ember = window.Ember;
-var EmberObject = Ember.Object;
+const Ember = window.Ember;
+const { Object } = Ember;
```
`Object: EmberObject`

teddyzeenny commented on ember_debug/view-debug.js:
```diff
@@ -1,18 +1,16 @@
-var Ember = window.Ember;
-var guidFor = Ember.guidFor;
+const Ember = window.Ember;
+const { guidFor, $, computed, run, Object, View, typeOf, Component, ViewUtils, A, keys } = Ember;
```
`Object: EmberObject`

teddyzeenny commented on ember_debug/view-debug.js:
```diff
@@ -1,18 +1,16 @@
+    highlightedElement;
 
-var ViewDebug = Ember.Object.extend(PortMixin, {
+const ViewDebug = Object.extend(PortMixin, {
```
`export default`

teddyzeenny commented on ember_debug/view-debug.js:
```diff
@@ -122,57 +119,57 @@ var ViewDebug = Ember.Object.extend(PortMixin, {
+        this.highlightView(viewElem, true);
       }
     })
     .on('mousedown.inspect-' + this.get('eventNamespace'), function() {
```
`, () => {`

teddyzeenny commented on ember_debug/view-debug.js:
```diff
@@ -192,7 +189,7 @@ var ViewDebug = Ember.Object.extend(PortMixin, {
     function pinView() {
```
We can remove `self`:
`const pinView = () =>`

teddyzeenny commented on ember_debug/view-debug.js:
```diff
@@ -297,27 +291,28 @@ var ViewDebug = Ember.Object.extend(PortMixin, {
-        viewClass = shortViewName(view), name;
+  inspectView(view, retained) {
+    const templateName = view.get('templateName') || view.get('_debugTemplateName'),
+          viewClass = shortViewName(view);
```
```js
let templateName = view.get('templateName') || view.get('_debugTemplateName');
let viewClass = shortViewName(view);
```

teddyzeenny commented on ember_debug/view-debug.js:
```diff
@@ -326,19 +321,19 @@ var ViewDebug = Ember.Object.extend(PortMixin, {
+      let model = this.modelForView(view);
       if (model) {
-        if (Ember.Object.detectInstance(model) || Ember.typeOf(model) === 'array') {
+        if (Object.detectInstance(model) || typeOf(model) === 'array') {
```
`EmberObject`

teddyzeenny commented on ember_debug/view-debug.js:
```diff
@@ -464,10 +458,9 @@ var ViewDebug = Ember.Object.extend(PortMixin, {
-    var self = this;
+  _highlightRange(rect, options) {
+    let div,
+        isPreview = options.isPreview;
```
```js
let div;
let isPreview = options.isPreview;
```

teddyzeenny commented:
@Mawaheb this is great!

> I couldn't refactor the contents of /vendor, when I tried, I got an error "Can't use block-scoped let, const out of restricted mode."

Yeah vendor files are not transpiled. We can't (and shouldn't) refactor them.

> Because arrow functions should always be anonymous, I couldn't/figure out how to refactor this one to use this instead of self.

You can declare them as variables instead of functions:
```js
const pinView = () => {
};
Ember.deprecate = (message, test, options) => {
};
```

Mawaheb commented:
Thank you for your feedback, I modified the code accordingly.

One thing, When I pulled the latest master, and when I rebased, there were some merge/rebase conflicts, I hope I fixed them the right way.

teddyzeenny commented:
Sorry @Mawaheb this needs another rebase.

Mawaheb commented:
Sorry for the inconvenience, I hope I got it right this time @teddyzeenny  :)

teddyzeenny commented on ember_debug/view-debug.js:
```diff
@@ -122,59 +119,59 @@ var ViewDebug = Ember.Object.extend(PortMixin, {
     $(layerDiv).remove();
     $(previewDiv).remove();
     this.get('_lastNodes').clear();
+    View.removeMutationListener(this.viewTreeChanged);
```
This should be removed.

teddyzeenny commented on ember_debug/view-debug.js:
```diff
@@ -226,57 +223,56 @@ var ViewDebug = Ember.Object.extend(PortMixin, {
+      this.hideLayer();
     };
+
+    View.addMutationListener(this.viewTreeChanged);
```
This line should be removed.

teddyzeenny commented on ember_debug/view-debug.js:
```diff
@@ -226,57 +223,56 @@ var ViewDebug = Ember.Object.extend(PortMixin, {
-    var applicationViewId = $(emberApp.rootElement).find('> .ember-view').attr('id');
-    var rootView = this.get('viewRegistry')[applicationViewId];
+    const applicationViewId = $(emberApp.rootElement).find('> .ember-view').attr('id');
+    let rootView = View.views[applicationViewId];
```
Replace with `let rootView = this.get('viewRegistry')[applicationViewId];`

teddyzeenny commented on ember_debug/view-debug.js:
```diff
@@ -399,13 +395,13 @@ var ViewDebug = Ember.Object.extend(PortMixin, {
       view = element;
     } else {
-      view = this.get('viewRegistry')[element.id];
+      view = View.views[element.id];
```
`view = this.get('viewRegistry')[element.id];`

teddyzeenny commented on ember_debug/view-debug.js:
```diff
@@ -603,39 +598,34 @@ var ViewDebug = Ember.Object.extend(PortMixin, {
-    return id && !this.get('viewRegistry')[id].get('_childViews');
+  _isGlimmer() {
+    let id = keys(View.views)[0];
+    return id && !View.views[id].get('_childViews');
```
```js
let id = Ember.keys(this.get('viewRegistry'))[0];
return id && !this.get('viewRegistry')[id].get('_childViews');
```

teddyzeenny commented on ember_debug/view-debug.js:
```diff
@@ -603,39 +598,34 @@ var ViewDebug = Ember.Object.extend(PortMixin, {
 
-  viewRegistry: computed('application', function() {
-    return this.get('application.__container__').lookup('-view-registry:main') || Ember.View.views;
-  }),
```
the above CP (`viewRegistry`) should not be deleted

Mawaheb commented:
I made the changes you mentioned, Please let me know if there are any further changes that should be made.

teddyzeenny commented:
Thanks @Mawaheb!

Mawaheb commented:
my pleasure :) :+1:
